File: utils/scripts/setup/filters/hb-filter.py
#!/usr/bin/env python3
import os,sys
W=int(os.environ.get("HB_FILTER_W","3500")); LOGF=os.environ.get("HB_FILTER_LOGF","")
PFX=os.environ.get("HB_FILTER_PFX","[HBWRAP"); SFX=os.environ.get("HB_FILTER_SFX","]")
BIN=os.environ.get("HB_FILTER_BIN","[HBBIN"); FLUSH=int(os.environ.get("HB_FILTER_FLUSH",str(4*W)))
logfh=open(LOGF,"ab",buffering=0) if LOGF else None
ALLOWED={9,10,13,27}
def bad(b): return any(((x not in ALLOWED) and (x<32 or x==127)) for x in b)
def chunks(b,nl):
  L=len(b)
  if L<=W: sys.stdout.buffer.write(b+nl); return
  n=(L+W-1)//W; s=0
  for i in range(1,n+1):
    e=min(s+W,L); chunk=b[s:e]; a, c = s+1, e
    sys.stdout.buffer.write((f"{PFX} {i}/{n} {a}..{c}{SFX} ").encode()+chunk+b"\n"); s=e
  if nl: sys.stdout.buffer.write(nl)
def soft(buf):
  while len(buf)>FLUSH:
    win=bytes(buf[:W]); del buf[:W]
    sys.stdout.buffer.write((f"{PFX} 1/1 1..{len(win)}{SFX} ").encode()+win+b"\n")
buf=bytearray()
while True:
  ch=sys.stdin.buffer.read(8192)
  if not ch: break
  buf.extend(ch); soft(buf)
  while True:
    pn=buf.find(b"\n"); pr=buf.find(b"\r")
    if pn<0 and pr<0: break
    if pn<0: pn=1<<30
    if pr<0: pr=1<<30
    pos = pn if pn<pr else pr
    delim=buf[pos:pos+1]; line=bytes(buf[:pos]); take=1
    if delim==b"\r" and buf[pos:pos+2]==b"\r\n": take=2
    raw=bytes(buf[:pos+take]); del buf[:pos+take]
    if logfh: logfh.write(raw)
    if bad(line): sys.stdout.buffer.write((f"{BIN} suppressed {len(line)} bytes\n").encode())
    else: chunks(line,b"\n")
if buf:
  if logfh: logfh.write(bytes(buf))
  if bad(bytes(buf)): sys.stdout.buffer.write((f"{BIN} suppressed {len(buf)} bytes").encode())
  else: chunks(bytes(buf),b"")

File: utils/scripts/setup/filters/hb-filter.pl
#!/usr/bin/env perl
use strict; use warnings; binmode STDIN; binmode STDOUT;
my $W=$ENV{HB_FILTER_W}||3500; my $LOGF=$ENV{HB_FILTER_LOGF}||"";
my $PFX=$ENV{HB_FILTER_PFX}||"[HBWRAP"; my $SFX=$ENV{HB_FILTER_SFX}||"]";
my $BIN=$ENV{HB_FILTER_BIN}||"[HBBIN"; my $FLUSH=$ENV{HB_FILTER_FLUSH}||(4*$W);
my $logfh; if ($LOGF ne ""){ open($logfh,">>",$LOGF) or die $!; binmode $logfh; }
sub bad{ my($s)=@_; return ($s=~/[\x00-\x08\x0B\x0C\x0E-\x1A\x1C-\x1F\x7F]/)?1:0; }
sub chunks{ my($line,$nl)=@_; my $L=length($line); if($L<=$W){ print STDOUT $line,$nl; return; }
  my $n=int(($L+$W-1)/$W); my $start=0; for(my $i=1;$i<=$n;$i++){ my $end=$start+$W; $end=$L if $end>$L;
    my $chunk=substr($line,$start,$end-$start); my $a=$start+1; my $b=$end;
    print STDOUT "$PFX $i/$n $a..$b$SFX $chunk\n"; $start=$end; } print STDOUT $nl if $nl ne ""; }
sub soft{ my($r)=@_; my $buf=$$r; while(length($buf)>$FLUSH){ my $w=substr($buf,0,$W,"");
    print STDOUT "$PFX 1/1 1..".length($w)."$SFX $w\n"; } $$r=$buf; }
my $buf=""; while(1){ my $n=sysread(STDIN,my $c,8192); last if !defined($n)||$n==0; $buf.=$c; soft(\$buf);
  while($buf=~/[\r\n]/){ my $pn=index($buf,"\n"); $pn=1<<30 if $pn<0; my $pr=index($buf,"\r"); $pr=1<<30 if $pr<0;
    my $pos=($pn<$pr)?$pn:$pr; my $d=substr($buf,$pos,1); my $line=substr($buf,0,$pos);
    my $take=1; $take=2 if $d eq "\r" && substr($buf,$pos+1,1) eq "\n";
    my $raw=substr($buf,0,$pos+$take); $buf=substr($buf,$pos+$take); print $logfh $raw if $logfh;
    my $nl="\n"; if(bad($line)){ print STDOUT "$BIN suppressed ".length($line)." bytes$nl"; } else { chunks($line,$nl); } } }
if(length($buf)){ print $logfh $buf if $logfh; if(bad($buf)){ print STDOUT "$BIN suppressed ".length($buf)." bytes"; } else { chunks($buf,""); } }
